File: src/python/main/social_relationship.py
from time import time

from redis import Redis


def following_key(user):
    return 'following:' + user


def followers_key(user):
    return 'followers:' + user


def common_following_key(user1, user2):
    return 'common:following:{}:{}'.format(user1, user2)


class SocialRelationship:
    def __init__(self, client: Redis, user):
        self.client = client
        self.user = user

    def follow(self, target):
        """关注目标用户"""
        self.client.zadd(following_key(self.user), {target: time()})
        self.client.zadd(followers_key(target), {self.user: time()})

    def unfollow(self, target):
        """取关目标用户"""
        self.client.zrem(following_key(self.user), target)
        self.client.zrem(followers_key(target), self.user)

    def is_following(self, target):
        """判断是否关注着目标用户"""
        return self.client.zrank(following_key(self.user), target) is not None

    def get_all_following(self):
        """获取当前用户关注的所有人，并按最近关注时间倒序排列"""
        return self.client.zrevrange(following_key(self.user), 0, -1)

    def get_all_followers(self):
        """获取当前用户的所有粉丝，并按最近关注时间倒序排列"""
        return self.client.zrevrange(followers_key(self.user), 0, -1)

    def count_following(self):
        """获取当前用户关注的人数"""
        return self.client.zcard(following_key(self.user))

    def count_followers(self):
        """获取当前用户的粉丝数"""
        return self.client.zcard(followers_key(self.user))

    def get_common_following(self, one):
        """获取与某用户的共同关注"""
        common_key = common_following_key(self.user, one)
        self.client.zinterstore(
            common_key, (following_key(self.user), following_key(one)))
        return self.client.zrevrange(common_key, 0, -1)

File: java/src/main/java/SocialRelationship.java
import redis.clients.jedis.Jedis;
import utils.JedisUtils;

import java.util.Set;

public class SocialRelationship {
    private Jedis client = JedisUtils.getClient();
    private String user;

    public SocialRelationship(String user) {
        this.user = user;
    }

    /**
     * 关注目标用户
     *
     * @param target 用户
     */
    public void follow(String target) {
        String followingKey = JedisUtils.getFollowingKey(user);
        String followersKey = JedisUtils.getFollowersKey(target);
        long now = System.currentTimeMillis();
        client.zadd(followingKey, now, target);
        client.zadd(followersKey, now, user);
    }

    /**
     * 取关目标用户
     *
     * @param target 用户
     */
    public void unfollow(String target) {
        String followingKey = JedisUtils.getFollowingKey(user);
        String followersKey = JedisUtils.getFollowersKey(target);
        client.zrem(followingKey, target);
        client.zrem(followersKey, user);
    }

    /**
     * 判断是否关注着目标用户
     *
     * @param target 用户
     * @return 是否关注
     */
    public boolean isFollowing(String target) {
        String followingKey = JedisUtils.getFollowingKey(user);
        return client.zrank(followingKey, target) != null;

    }

    /**
     * 获取当前用户关注的所有人，并按最近关注时间倒序排列
     *
     * @return 关注人集合
     */
    public Set<String> getAllFollowing() {
        String followingKey = JedisUtils.getFollowingKey(user);
        return client.zrevrange(followingKey, 0, -1);
    }

    /**
     * 获取当前用户的所有粉丝，并按最近关注时间倒序排列
     *
     * @return 粉丝集合
     */
    public Set<String> getAllFollowers() {
        String followersKey = JedisUtils.getFollowersKey(user);
        return client.zrevrange(followersKey, 0, -1);
    }

    /**
     * 获取当前用户关注的人数
     *
     * @return 人数
     */
    public Long countFollowing() {
        String followingKey = JedisUtils.getFollowingKey(user);
        return client.zcard(followingKey);
    }

    /**
     * 获取当前用户的粉丝数
     *
     * @return 人数
     */
    public Long countFollowers() {
        String followersKey = JedisUtils.getFollowersKey(user);
        return client.zcard(followersKey);
    }

    /**
     * 获取与某用户的共同关注
     *
     * @param one 用户
     * @return 共同关注的用户集合
     */
    public Set<String> getCommonFollowing(String one) {
        String commonKey = JedisUtils.getCommonFollowingKey(user, one);
        client.zinterstore(commonKey, JedisUtils.getFollowingKey(user), JedisUtils.getFollowingKey(one));
        return client.zrevrange(commonKey, 0, -1);
    }
}